core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -991,6 +992,16 @@ public void deleteSnapshot(final Snapshot snapshot, final DeleteSnapshotListener
 
             @Override
             public ClusterState execute(ClusterState currentState) throws Exception {
+                RestoreInProgress restoreInProgress = currentState.custom(RestoreInProgress.TYPE);
+                if (restoreInProgress != null) {
+                    for (RestoreInProgress.Entry entry : restoreInProgress.entries()) {
+                        if (entry.snapshot().equals(snapshot)) {
```

Comment:
One race condition is possible: since we only keep around the current index generational file as well as the previous one (`index-{N}` and `index-{N-1}`), its possible that the restore operation reads what the current `N` value is, and before reading the `index-N` file, we execute two snapshot deletions in succession which updates the index gen file twice, so the only index gen files we now have in the repo are `index-{N+2}` and `index-{N+1}`, so when the restore operation goes to read `index-{N}`, its not there.  I will update this PR so that deletions are not allowed if *any* restore is taking place.